src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -0,0 +1,244 @@
+  std::string cache_dir = m_cache_root_dir;
+
+  int dir = m_dir_num - 1;
+  while (dir >= 0) {
+    efs::create_directories(cache_dir + "/" + std::to_string(dir));
+    dir --;
+  }
+  return 0;
+}
+
+int ObjectCacheStore::do_promote(std::string pool_name, std::string object_name) {
+  ldout(m_cct, 20) << "to promote object = "
+                   << object_name << " from pool: "
+                   << pool_name << dendl;
+
+  int ret = 0;
+  std::string cache_file_name =  pool_name + object_name;
+  {
+    Mutex::Locker _locker(m_ioctxs_lock);
+    if (m_ioctxs.find(pool_name) == m_ioctxs.end()) {
+      librados::IoCtx* io_ctx = new librados::IoCtx();
+      ret = m_rados->ioctx_create(pool_name.c_str(), *io_ctx);
+      if (ret < 0) {
+        lderr(m_cct) << "fail to create ioctx" << dendl;
+        return ret;
+      }
+      m_ioctxs.emplace(pool_name, io_ctx);
+    }
+  }
+
+  assert(m_ioctxs.find(pool_name) != m_ioctxs.end());
+
+  librados::IoCtx* ioctx = m_ioctxs[pool_name];
+
+  librados::bufferlist* read_buf = new librados::bufferlist();
+
+  auto ctx = new FunctionContext([this, read_buf, cache_file_name](int ret) {
+      handle_promote_callback(ret, read_buf, cache_file_name);
+   });
+
+   return promote_object(ioctx, object_name, read_buf, ctx);
+}
+
+int ObjectCacheStore::handle_promote_callback(int ret, bufferlist* read_buf,
+  std::string cache_file_name) {
+  ldout(m_cct, 20) << " cache_file_name: " << cache_file_name << dendl;
+
+  // rados read error
+  if(ret != -ENOENT && ret < 0) {
+    lderr(m_cct) << "fail to read from rados" << dendl;
+
+    m_policy->update_status(cache_file_name, OBJ_CACHE_NONE);
+    delete read_buf;
+    return ret;
+  }
+
+  if (ret == -ENOENT) {
+    // object is empty
+    ret = 0;
+  }
+
+  std::string cache_dir = "";
+  uint32_t file_size = ret;
+
+  if (m_dir_num > 0) {
+    auto const pos = cache_file_name.find_last_of('.');
+    cache_dir = "/" + std::to_string(stoul(cache_file_name.substr(pos+1)) % m_dir_num);
+  }
+  // write to cache
+  ObjectCacheFile cache_file(m_cct, cache_dir + "/" + cache_file_name);
+
+  ret = cache_file.write_object_to_file(*read_buf, file_size);
+  if (ret < 0) {
+    lderr(m_cct) << "fail to write cache file" << dendl;
+
+    m_policy->update_status(cache_file_name, OBJ_CACHE_NONE);
+    delete read_buf;
+    return ret;
+  }
+
+  // update metadata
+  assert(OBJ_CACHE_SKIP == m_policy->get_status(cache_file_name));
+  m_policy->update_status(cache_file_name, OBJ_CACHE_PROMOTED);
+  assert(OBJ_CACHE_PROMOTED == m_policy->get_status(cache_file_name));
+
+  delete read_buf;
+  return ret;
+
+  evict_objects();
+}
+
+int ObjectCacheStore::lookup_object(std::string pool_name,
+                                    std::string object_name) {
+  ldout(m_cct, 20) << "object name = " << object_name
+                   << " in pool: " << pool_name << dendl;
+
+  int pret = -1;
+  cache_status_t ret = m_policy->lookup_object(pool_name + object_name);
```

Comment:
I would add some special character delimiter (e.g. ':') between pool name (id) and object name. Consider pool/object examples: "rbd/1a" and "rbd1/a". They will collide in your case.

Reply:
Also, I think you need to include snap_id into the lookup key.